pkg/apis/build/v1alpha1/image_builds.go

```diff
@@ -106,6 +132,72 @@ func (im *Image) build(sourceResolver *SourceResolver, builder AbstractBuilder,
 	}
 }
 
+func (im *Image) rebase(builder AbstractBuilder, previousRunImage string, nextBuildNumber int64) *Build {
+	buildNumber := strconv.Itoa(int(nextBuildNumber))
+	return &Build{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    im.Namespace,
+			GenerateName: im.generateBuildName(buildNumber),
+			OwnerReferences: []metav1.OwnerReference{
+				*kmeta.NewControllerRef(im),
+			},
+			Labels: im.labels(map[string]string{
+				BuildNumberLabel: buildNumber,
+				ImageLabel:       im.Name,
+			}),
+			Annotations: map[string]string{
+				BuildReasonAnnotation: BuildReasonStack,
+			},
+		},
+		Spec: BuildSpec{
+			Tags:           im.generateTags(buildNumber),
+			ServiceAccount: im.Spec.ServiceAccount,
+			Rebase: &RebaseConfig{
+				PreviousRunImage: runImageConfig(builder, previousRunImage),
+				LatestRunImage:   runImageConfig(builder, builder.RunImage()),
+			},
+		},
+	}
+}
+
+func runImageConfig(builder AbstractBuilder, ref string) RunImageConfig {
+	secret := ""
+	if len(builder.ImageRef().ImagePullSecrets) > 0 {
+		secret = builder.ImageRef().ImagePullSecrets[0].Name
+	}
+	return RunImageConfig{
+		Ref:              ref,
+		BuilderNamespace: builder.GetObjectMeta().GetNamespace(),
+		PullSecretName:   secret,
+	}
+}
+
+type RunImageConfig struct {
+	Ref              string `json:"ref"`
+	BuilderNamespace string `json:"namespace"`
```

Comment:
Why do we need a builder namespace here?

Reply:
If it does use the imagePullSecret for the builder, it needs to know the namespace of the builder as that is the namespace where the pull secret will lie

Reply:
At this point we really should only support secrets in the same namespace as the build itself.

Reply:
I go a little back and forth on this. The run image feels like a builder concern but I also see the argument of it being a build concern. It lies in that weird grey area, where it is entirely conceivable that the build gets to decide its run image, but when the run image is not explicit, we implicitly rely on the builder secrets to pull it